sycl/include/sycl/ext/oneapi/experimental/graph.hpp

```diff
@@ -0,0 +1,255 @@
+  void make_edge(node &Src, node &Dest);
+
+  /// Finalize modifiable graph into an executable graph.
+  /// @param PropList Property list used to pass properties for finalization.
+  /// @return Executable graph object.
+  command_graph<graph_state::executable>
+  finalize(const property_list &PropList = {}) const;
+
+  /// Change the state of a queue to be recording and associate this graph with
+  /// it.
+  /// @param RecordingQueue The queue to change state on and associate this
+  /// graph with.
+  /// @return True if the queue had its state changed from executing to
+  /// recording.
+  bool begin_recording(queue &RecordingQueue);
+
+  /// Change the state of multiple queues to be recording and associate this
+  /// graph with each of them.
+  /// @param RecordingQueues The queues to change state on and associate this
+  /// graph with.
+  /// @return True if any queue had its state changed from executing to
+  /// recording.
+  bool begin_recording(const std::vector<queue> &RecordingQueues);
+
+  /// Set all queues currently recording to this graph to the executing state.
+  /// @return True if any queue had its state changed from recording to
+  /// executing.
+  bool end_recording();
+
+  /// Set a queues currently recording to this graph to the executing state.
+  /// @param RecordingQueue The queue to change state on.
+  /// @return True if the queue had its state changed from recording to
+  /// executing.
+  bool end_recording(queue &RecordingQueue);
+
+  /// Set multiple queues currently recording to this graph to the executing
+  /// state.
+  /// @param RecordingQueues The queues to change state on.
+  /// @return True if any queue had its state changed from recording to
+  /// executing.
+  bool end_recording(const std::vector<queue> &RecordingQueues);
+
+private:
+  /// Constructor used internally by the runtime.
+  /// @param Impl Detail implementation class to construct object with.
+  command_graph(const std::shared_ptr<detail::graph_impl> &Impl) : impl(Impl) {}
+
+  /// Template-less implementation of add() for CGF nodes.
+  /// @param CGF Command-group function to add.
+  /// @param Dep List of predecessor nodes.
+  /// @return Node added to the graph.
+  node add_impl(std::function<void(handler &)> CGF,
+                const std::vector<node> &Dep);
+
+  /// Template-less implementation of add() for empty nodes.
+  /// @param Dep List of predecessor nodes.
+  /// @return Node added to the graph.
+  node add_impl(const std::vector<node> &Dep);
+
+  template <class Obj>
+  friend decltype(Obj::impl)
+  sycl::detail::getSyclObjImpl(const Obj &SyclObject);
+  template <class T>
+  friend T sycl::detail::createSyclObjFromImpl(decltype(T::impl) ImplObj);
+
+  std::shared_ptr<detail::graph_impl> impl;
+};
+
+template <> class __SYCL_EXPORT command_graph<graph_state::executable> {
+public:
+  /// An executable command-graph is not user constructable.
+  command_graph() = delete;
+
+  /// Constructor used by internal runtime.
```

Comment:
Why the modifiable command_graph constructor is private, whereas the executable command_graph constructor is public?

Reply:
Addressed in https://github.com/reble/llvm/pull/219